File: solutions/typescript/155-ransom-note.ts
/**
 * Problem 155: Ransom Note (LeetCode 383)
 * Difficulty: Easy
 * Language: TypeScript
 */
function canConstruct(ransomNote: string, magazine: string): boolean {
  const freq = [...magazine].reduce<Record<string, number>>(
    (acc, ch) => ({ ...acc, [ch]: (acc[ch] ?? 0) + 1 }), {});
  return [...ransomNote].every(ch => {
    freq[ch] = (freq[ch] ?? 0) - 1;
    return freq[ch] >= 0;
  });
}

File: solutions/clojure/155-ransom-note.clj
;; Problem 155: Ransom Note (LeetCode 383)
;; Difficulty: Easy
;; Language: Clojure
;; 
(defn can-construct [ransom magazine]
  (let [freq (frequencies magazine)]
    (second (reduce (fn [[f ok] ch]
              (let [cnt (get f ch 0)]
                [(update f ch (fnil dec 0)) (and ok (pos? cnt))]))
            [freq true] ransom))))
